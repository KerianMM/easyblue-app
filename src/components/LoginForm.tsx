import React from 'react';
import {Button, Grid, IconButton, InputAdornment, Paper, TextField} from "@material-ui/core";
import {Visibility, VisibilityOff} from "@material-ui/icons";
import {ReactSVG} from "react-svg";
import {useRouter} from "next/router";
import {NextRouter} from "next/dist/next-server/lib/router/router";

import FormErrors from "./FormErrors";
import {UserService} from "../service/user";
import {FormErrorContainer} from "../interfaces/form/error";
import {LoginFormDatas, LoginState} from "../interfaces/form/login";
import {LoginResponse} from "../interfaces/response/login";
import {EmailValidator} from "../utils/validators/email";

const LoginForm: React.FC = () => {
    const router: NextRouter = useRouter();

    const [values, setValues] = React.useState<LoginState>({
        email: {
            value: '',
            error: false,
            helper: null
        },
        password: {
            value: '',
            error: false,
            helper: null,
            show: false
        },
    });

    const [formErrors, setFormErrors] = React.useState<FormErrorContainer>({errors: []});

    /**
     * Dispatch SetStateAction to add an error message
     */
    const addFormError = () => {
        setFormErrors({
            errors: [
                ...formErrors.errors,
                {message: 'Connexion impossible... Veuillez réessayer plus tard.'}
            ]
        });
    };

    /**
     * Dispatch SetStateAction to update fields values
     * Clear FormErrors
     */
    const handleChange = (prop: keyof LoginState) => (event: React.ChangeEvent<HTMLInputElement>) => {
        setValues({
            ...values,
            [prop]: {
                ...values[prop],
                value: event.target.value,
                error: false,
                helper: ''
            }
        });

        if (formErrors.errors.length) {
            setFormErrors({errors: []});
        }
    };

    /**
     * Dispatch SetStateAction to display or hide password value
     */
    const handleClickShowPassword = () => {
        setValues({...values, password: {...values.password, show: !values.password.show}});
    };

    /**
     * Dispatch SetStateAction to ignore password visibility click
     */
    const handleMouseDownPassword = (event: React.MouseEvent<HTMLButtonElement>) => {
        event.preventDefault();
    };

    /**
     * Dispatch SetStateAction to add form field error
     */
    const handleInvalidField = (prop: keyof LoginState, helper: string) => {
        setValues({
            ...values,
            [prop]: {
                ...values[prop],
                error: true,
                helper: helper
            }
        });
    };

    /**
     * Log user and redirect to dashboard or add FormError
     */
    const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        const formData: LoginFormDatas = {
            email: values.email.value,
            password: values.password.value
        };

        if (!formData.email.length) {
            handleInvalidField('email', 'Champs requis');
        } else if (!formData.password.length) {
            handleInvalidField('password', 'Champs requis');
        } else if (!EmailValidator.isValid(formData.email)) {
            handleInvalidField('email', 'Format invalide');
        } else {
            try {
                const response: Response = await UserService.fetchLogin(formData);

                const datas: LoginResponse = await response.json();

                if (datas.valid) {
                    sessionStorage.setItem('email', formData.email);
                    router.push('/dashboard');
                } else if (typeof datas.field === "string" && typeof datas.error === "string") {
                    handleInvalidField((datas.field === 'email' ? 'email' : 'password'), datas.error);
                } else {
                    addFormError();
                }
            } catch (e) {
                addFormError();
            }
        }
    };

    return (
        <>
            <Paper id="login-card">
                <form onSubmit={handleSubmit}>
                    <Grid>
                        <div className="logo">
                            <ReactSVG src="/img/icons/logo.svg"/>
                        </div>
                    </Grid>

                    <Grid container id="login-form">
                        <FormErrors formErrors={formErrors}/>

                        <Grid container className="raw">
                            <TextField
                                label="Adresse e-mail"
                                error={values.email.error}
                                helperText={values.email.helper}
                                fullWidth
                                value={values.email.value}
                                onChange={handleChange('email')}
                            />
                        </Grid>
                        <Grid container className="raw">
                            <TextField
                                label="Mot de passe"
                                type={values.password.show ? 'text' : 'password'}
                                value={values.password.value}
                                error={values.password.error}
                                helperText={values.password.helper}
                                fullWidth
                                onChange={handleChange('password')}
                                InputProps={{
                                    endAdornment: (
                                        <InputAdornment position="end">
                                            <IconButton
                                                aria-label="toggle password visibility"
                                                onClick={handleClickShowPassword}
                                                onMouseDown={handleMouseDownPassword}
                                            >
                                                {values.password.show ? <Visibility/> : <VisibilityOff/>}
                                            </IconButton>
                                        </InputAdornment>
                                    ),
                                }}
                            />
                        </Grid>
                        <Grid container className="raw">
                            <div className="m-auto">
                                <Button size="small" type="submit">Connexion</Button>
                            </div>
                        </Grid>
                        <Grid container className="raw">
                            <div className="m-auto">
                                <Button size="small">Mot de passe oublié</Button>
                            </div>
                        </Grid>
                    </Grid>
                </form>
            </Paper>
        </>
    );
};

export default LoginForm;
