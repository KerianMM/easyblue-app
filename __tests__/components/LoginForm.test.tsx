import '@testing-library/jest-dom';
import React from 'react';
import renderer from 'react-test-renderer';
import LoginForm from "../../src/components/LoginForm";

test('<LoginForm/> SNAPSHOT', () => {
    const tree = renderer.create(<LoginForm/>).toJSON();

    expect(tree).toMatchSnapshot();
});
