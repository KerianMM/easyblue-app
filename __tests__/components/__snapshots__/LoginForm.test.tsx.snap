// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm/> SNAPSHOT 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  id="login-card"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiGrid-root"
    >
      <div
        className="logo"
      >
        <div />
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-container"
      id="login-form"
    >
      <div
        className="MuiGrid-root raw MuiGrid-container"
      />
      <div
        className="MuiGrid-root raw MuiGrid-container"
      >
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            Adresse e-mail
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root raw MuiGrid-container"
      >
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            Mot de passe
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="password"
              value=""
            />
            <div
              className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <button
                aria-label="toggle password visibility"
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root raw MuiGrid-container"
      >
        <div
          className="m-auto"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="submit"
          >
            <span
              className="MuiButton-label"
            >
              Connexion
            </span>
          </button>
        </div>
      </div>
      <div
        className="MuiGrid-root raw MuiGrid-container"
      >
        <div
          className="m-auto"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Mot de passe oubli√©
            </span>
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
